sklearn/cluster/optics_.py

```diff
@@ -817,3 +997,397 @@ def _get_leaves(node, arr):
+        The start of the downward region.
+
+    ixi: float, between 0 and 1
+        The inverse xi, i.e. `1 - xi`
+
+    min_samples : integer
+       The same as the min_samples given to OPTICS. Up and down steep
+       regions can't have more then `min_samples` consecutive non-steep
+       points.
+
+    n_samples : integer
+        Total number of samples.
+
+    Returns
+    -------
+    index : integer
+        The current index iterating over all the samples.
+
+    end : integer
+        The end of the downward region, which can be behind the index.
+    """
+    # print("extend down start")
+    non_downward_points = 0
+    index = start
+    end = start
+    # find a maximal downward area
+    while index + 1 < n_samples:
+        # print("index", index)
+        # print("r", reachability_plot[index], "r + 1",
+        #       reachability_plot[index + 1])
+        index += 1
+        if _steep_downward(reachability_plot, index, ixi):
+            # print("steep")
+            non_downward_points = 0
+            end = index + 1
+        elif reachability_plot[index] >= reachability_plot[index + 1]:
+            # print("just down")
+            # it's not a steep downward point, but still goes down.
+            non_downward_points += 1
+            # region should include no more than min_samples consecutive
+            # non downward points.
+            if non_downward_points == min_samples:
+                # print("non downward")
+                break
+        else:
+            break
+    # print("extend end")
+    return index, end
+
+
+def _extend_upward(reachability_plot, start, ixi, min_samples, n_samples):
+    """Extend the upward area until it's maximal.
+
+    Parameters
+    ----------
+    reachability_plot : array, shape (n_samples)
+        The reachability plot, i.e. reachability ordered according to
+        the calculated ordering, all computed by OPTICS.
+
+    start : integer
+        The start of the upward region.
+
+    ixi: float, between 0 and 1
+        The inverse xi, i.e. `1 - xi`
+
+    min_samples : integer
+       The same as the min_samples given to OPTICS. Up and down steep
+       regions can't have more then `min_samples` consecutive non-steep
+       points.
+
+    n_samples : integer
+        Total number of samples.
+
+    Returns
+    -------
+    index : integer
+        The current index iterating over all the samples.
+
+    end : integer
+        The end of the upward region, which can be behind the index.
+    """
+    # print("extend up start")
+    non_upward_points = 0
+    index = start
+    end = start
+    # find a maximal upward area
+    while index + 1 < n_samples:
+        # print("index", index)
+        # print("r", reachability_plot[index], "r + 1",
+        #       reachability_plot[index + 1])
+        index += 1
+        if _steep_upward(reachability_plot, index, ixi):
+            # print("steep")
+            non_upward_points = 0
+            end = index + 1
```

Comment:
If `end` indicates the element *after* the span of interest, maybe we should adopt the Python convention of calling this index `stop` rather than `end`.